pkg/pubsub/transformer/transformer.go

```diff
@@ -0,0 +1,147 @@
+	Message *PubSubMessage `json:"message,omitempty"`
+}
+
+// PubSubMessage matches the inner message format used by Push Subscriptions.
+type PubSubMessage struct {
+	// ID identifies this message. This ID is assigned by the server and is
+	// populated for Messages obtained from a subscription.
+	// This field is read-only.
+	ID string `json:"id,omitempty"`
+
+	// Data is the actual data in the message.
+	Data interface{} `json:"data,omitempty"`
+
+	// Attributes represents the key-value pairs the current message
+	// is labelled with.
+	Attributes map[string]string `json:"attributes,omitempty"`
+
+	// The time at which the message was published. This is populated by the
+	// server for Messages obtained from a subscription.
+	// This field is read-only.
+	PublishTime time.Time `json:"publish_time,omitempty"`
+}
+
+// Transformer implements the Pub/Sub transformer to convert generic Cloud
+// Pub/Sub events into compatible formats.
+type Transformer struct {
+	// client is the cloudevents client to use to receive events.
+	client cloudevents.Client
+}
+
+// Start starts the adapter. Note: Only call once, not thread safe.
+func (a *Transformer) Start(ctx context.Context) error {
+	var err error
+
+	// Receive Events on Pub/Sub.
+	if a.client == nil {
+		if a.client, err = kncloudevents.NewDefaultClient(); err != nil {
+			return fmt.Errorf("failed to create cloudevent client: %s", err.Error())
+		}
+	}
+
+	return a.client.StartReceiver(ctx, a.receive)
+}
+
+func (a *Transformer) receive(ctx context.Context, event cloudevents.Event, resp *cloudevents.EventResponse) error {
+	logger := logging.FromContext(ctx).With(zap.Any("event.id", event.ID()))
+
+	if event.Type() != v1alpha1.PubSubEventType {
+		logger.Warnw("Not attempting to convert non-Pub/Sub type.", zap.String("type", event.Type()))
+		resp.RespondWith(http.StatusNotModified, &event)
+		return nil
+	}
+
+	// Convert data to PushMessage.
+
+	push := cloudevents.NewEvent(event.SpecVersion())
+	push.Context = event.Context.Clone()
+
+	// This is a workaround with the current CloudEvents SDK, it does not
+	// handle extensions very well. The following code converts the
+	// "attributes" extension into a attributes map that Pub/Sub push
+	// consumers expect.
+	attrs := make(map[string]string, 0)
+	if at, ok := event.Extensions()["attributes"]; ok {
+		if m, ok := at.(map[string]interface{}); ok {
+			for k, v := range m {
+				if vs, ok := v.([]string); ok {
+					var err error
+					attrs[k], err = strconv.Unquote(vs[0])
+					if err != nil {
+						logger.Warnw("Failed to unquote attributes string value.", zap.Error(err))
+					}
+				}
+			}
+		}
+	}
+
+	msg := &PubSubMessage{
+		ID:          event.ID(),
+		Attributes:  attrs,
+		PublishTime: event.Time(),
+	}
+
+	var subscription string
+	if sub, ok := event.Extensions()["subscription"]; ok {
+		subscription = sub.(string)
+	}
+
+	var raw json.RawMessage
+	if err := event.DataAs(&raw); err != nil {
```

Comment:
Does this mean that we sometimes decode as JSON and sometimes as binary for the same `spec`?

I'm concerned that this could corrupt binary payloads which happen to parse as JSON, e.g. `  "word"  ` if the spaces are actually important.

Reply:
I tested this case and it works fine.